packages/cdk-assets/lib/publishing.ts

```diff
@@ -0,0 +1,114 @@
+import { AssetManifest, IManifestEntry } from './asset-manifest';
+import { IAws } from "./aws-operations";
+import { makeAssetHandler } from "./handlers";
+
+export interface IPublishProgressListener {
+  onAssetStart(event: IPublishProgress): void;
+  onAssetEnd(event: IPublishProgress): void;
+  onEvent(event: IPublishProgress): void;
+  onError(event: IPublishProgress): void;
+}
+
+export interface IPublishProgress {
+  readonly message: string;
+  readonly currentAsset?: IManifestEntry;
+  readonly percentComplete: number;
+
+  /**
+   * Abort the current publishing operation
+   */
+  abort(): void;
+}
+
+export interface AssetPublishingOptions {
+  /**
+   * Manifest containing assets to publish
+   */
+  readonly manifest: AssetManifest;
+
+  /**
+   * Entry point for AWS client
+   */
+  readonly aws: IAws;
+
+  /**
+   * Listener for progress events
+   *
+   * @default No listener
+   */
+  readonly progressListener?: IPublishProgressListener;
+
+  /**
+   * Whether to throw at the end if there were errors
+   *
+   * @default true
+   */
+  readonly throwOnError?: boolean;
+}
+
+export class AssetPublishing implements IPublishProgress {
+  public message: string = 'Starting';
+  public currentAsset?: IManifestEntry;
+  public readonly failedAssets = new Array<IManifestEntry>();
+  private readonly assets: IManifestEntry[];
+  private readonly errors = new Array<Error>();
+
+  private readonly totalOperations: number;
+  private completedOperations: number = 0;
+  private aborted = false;
+
+  constructor(private readonly options: AssetPublishingOptions) {
+    this.assets = this.options.manifest.entries;
+    this.totalOperations = this.assets.length;
+  }
+
+  public async publish(): Promise<void> {
+    for (const asset of this.assets) {
+      if (this.aborted) { break; }
+      this.currentAsset = asset;
+
+      try {
+        if (this.progress('onAssetStart', `Packaging ${asset.id}`)) { break; }
+
+        const handler = makeAssetHandler(this.options.manifest, asset, this.options.aws, m => this.progress('onEvent', m));
+        await handler.publish();
+
+        this.completedOperations++;
+        if (this.progress('onAssetEnd', `Published ${asset.id}`)) { break; }
+      } catch (e) {
+        this.errors.push(e);
+        this.failedAssets.push(asset);
```

Comment:
Feels like we should bundle `asset` and `e` together in the array so they can be correlated (or even do something like `e.asset = asset` so that it will be possible to find the asset from the error.